### PR TITLE
ComboBox: Only allow focus to be put back on input if focus is still inside the ComboBox

### PR DESCRIPTION
**Pull request by chang47**
#### Pull request checklist

- [#5595] Addresses an existing issue: Fixes #0000
- [X] Include a change request file using `$ npm run change`

#### Description of changes

Currently the ComboBox is too aggressive in how it sets focus back into itself. There are 2 points where the ComboBox puts focus back into itself and that's in:
1) Whenever we update the prop and state in the ComboBox in ComponentDidUpdate
2) Whenever we dismiss the callout in dismiss()

This works for most cases, however there are situations where the user would click out of the ComboBox which will cause the ComboBox to change. As a result putting focus back on to itself in the 2 ways mentioned above while focus is also being changed to somewhere outside of the ComboBox.

As a result, we now have 2 locations that appear to have focus, the ComboBox and the external location.

To fix this problem, I added checks to ensure that we should only put focus back into the input element in the ComboBox if our current focus is still somewhere in the ComboBox component.

#### Focus areas to test

ComboBox page.

To test my changes I:
1. Added a Checkbox to the ComboBox Page similar to what #5595 did and validated that the problem no longer repros
2. Tested in my own external app that the focus related problem we focus something outside of the ComboBox we only have focus on the combo box now.

### CODE REVIEW

File: packages/office-ui-fabric-react/src/components/ComboBox/ComboBox.tsx [hunk 1]
```diff
@@ -232,13 +234,14 @@ export class ComboBox extends BaseComponent<IComboBoxProps, IComboBoxState> {
...
-        focused &&
-        this._autofill.current &&
-        document.activeElement !== this._autofill.current.inputElement)
+      this._focusInputAfterBlur &&
+      (isOpen ||
+        (prevState.isOpen &&
+          !isOpen &&
+          this._focusInputAfterClose &&
+          focused &&
+          this._autofill.current &&
```
Comment (jspurlin) [inline packages/office-ui-fabric-react/src/components/ComboBox/ComboBox.tsx]: This if check (and below) are getting unruly, consider pulling them into their own helper functions with good comments on the different sets of conditions

File: packages/office-ui-fabric-react/src/components/ComboBox/ComboBox.tsx [hunk 2]
```diff
@@ -1322,16 +1329,12 @@ export class ComboBox extends BaseComponent<IComboBoxProps, IComboBoxState> {
    * Handles dismissing (cancelling) the menu
    */
   private _onDismiss = (): void => {
-    // reset the selected index
-    // to the last valud state
-    this._resetSelectedIndex();
-
     // close the menu and focus the input
-    this.setState({ isOpen: false });
+    this._setOpenStateAndFocusOnClose(false, false);
```
Comment (jspurlin) [inline packages/office-ui-fabric-react/src/components/ComboBox/ComboBox.tsx]: >false, fals [](start = 38, length = 11)

Can you add inline comments here?

File: packages/office-ui-fabric-react/src/components/ComboBox/ComboBox.tsx [hunk 3]
```diff
@@ -1322,16 +1329,12 @@ export class ComboBox extends BaseComponent<IComboBoxProps, IComboBoxState> {
...
 
-    if (this._autofill.current && this._focusInputAfterClose) {
-      this._autofill.current.focus();
-    }
```
Comment (jspurlin) [inline packages/office-ui-fabric-react/src/components/ComboBox/ComboBox.tsx]: Are there any cases when we want to focus the input after close? Now you are always sending false here

File: packages/office-ui-fabric-react/src/components/ComboBox/ComboBox.tsx [hunk 4]
```diff
@@ -142,6 +142,7 @@ export class ComboBox extends BaseComponent<IComboBoxProps, IComboBoxState> {
   private _processingTouch: boolean;
   private _lastTouchTimeoutId: number | undefined;
   private _processingExpandCollapseKeyOnly: boolean;
+  private _focusInputAfterBlur: boolean;
```
Comment (jspurlin) [inline packages/office-ui-fabric-react/src/components/ComboBox/ComboBox.tsx]: Seems like this variable mimics this.state.focused or if focusInputAfterClose is false.  It seems like there's a lot of variables tracking more or less the same thing, is there any way we can simplify this?

File: packages/office-ui-fabric-react/src/components/ComboBox/ComboBox.tsx (hunk 2 above)
Comment (chang47) [reply packages/office-ui-fabric-react/src/components/ComboBox/ComboBox.tsx]: Will do!

File: packages/office-ui-fabric-react/src/components/ComboBox/ComboBox.tsx (hunk 1 above)
Comment (chang47) [reply packages/office-ui-fabric-react/src/components/ComboBox/ComboBox.tsx]: Will do, I was actually trying to figure out how to get rid of _focusInputAfterBlur and use focused instead.

File: packages/office-ui-fabric-react/src/components/ComboBox/ComboBox.tsx (hunk 4 above)
Comment (chang47) [reply packages/office-ui-fabric-react/src/components/ComboBox/ComboBox.tsx]: I actually think we can use the focused state to mimic this and am currently looking to see how I can use that.

File: packages/office-ui-fabric-react/src/components/ComboBox/ComboBox.tsx (hunk 3 above)
Comment (chang47) [reply packages/office-ui-fabric-react/src/components/ComboBox/ComboBox.tsx]: onDismiss is only ever triggered when we click out via a click event. Any other event was caught by another event handler. 

So with all that in mind, there shouldn't be any scenario that should put focus in the autofill when we click out of it.

File: packages/office-ui-fabric-react/src/components/ComboBox/ComboBox.tsx (hunk 2 above)
Comment (jspurlin) [reply packages/office-ui-fabric-react/src/components/ComboBox/ComboBox.tsx]: NIT: inline comments on the static booleans being passed to setOpenStateAndFocusOnClose

---
In reply to: [206741269](https://github.com/OfficeDev/office-ui-fabric-react/pull/5663#discussion_r206741269) [](ancestors = 206741269)

**Pull request merged**